File: src/embed/llamacpp.ts
import { config } from '../config.js';
import { logger } from '../logger.js';

interface OAEmbedResponse {
  data: Array<{ embedding: number[]; index: number }>;
}

export async function embedLlamaCpp(texts: string[], batchSize = 8): Promise<number[][]> {
  const result: number[][] = [];
  for (let i = 0; i < texts.length; i += batchSize) {
    const batch = texts.slice(i, i + batchSize);
    const res = await fetch(`${config.LLAMACPP_EMBED_URL}/embeddings`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ model: config.LLAMACPP_EMBED_MODEL, input: batch }),
    });
    if (!res.ok) throw new Error(`llamacpp embed failed (${res.status}): ${await res.text()}`);
    const json = await res.json() as OAEmbedResponse;
    json.data.sort((a, b) => a.index - b.index);
    for (const d of json.data) {
      if (d.embedding.length !== config.EMBED_DIM) throw new Error(`Dim mismatch: ${d.embedding.length}`);
      result.push(d.embedding);
    }
    logger.debug({ done: result.length, total: texts.length }, 'embed (llamacpp)');
  }
  return result;
}

File: crates/embed/src/llamacpp.rs
use rag_common::{AppError, Config, Result};
use serde::Deserialize;
use serde_json::json;

const BATCH: usize = 8;

#[derive(Debug, Deserialize)]
struct EmbedResponse {
    data: Vec<EmbedItem>,
}

#[derive(Debug, Deserialize)]
struct EmbedItem {
    index: usize,
    embedding: Vec<f32>,
}

pub async fn embed(texts: &[String]) -> Result<Vec<Vec<f32>>> {
    let cfg = Config::get();
    let client = reqwest::Client::new();
    let url = format!(
        "{}/embeddings",
        cfg.llamacpp_embed_url.trim_end_matches('/')
    );

    let mut out: Vec<Vec<f32>> = Vec::with_capacity(texts.len());
    for chunk in texts.chunks(BATCH) {
        let body = json!({
            "model": cfg.llamacpp_embed_model,
            "input": chunk,
        });
        let resp = client
            .post(&url)
            .json(&body)
            .send()
            .await?
            .error_for_status()
            .map_err(AppError::http)?;
        let mut parsed: EmbedResponse = resp.json().await?;
        parsed.data.sort_by_key(|d| d.index);
        for item in parsed.data {
            if item.embedding.len() != cfg.embed_dim as usize {
                return Err(AppError::DimMismatch {
                    expected: cfg.embed_dim as usize,
                    got: item.embedding.len(),
                });
            }
            out.push(item.embedding);
        }
    }
    Ok(out)
}
